File: Systems/SecuritySystem.py
import logging
import os
from configparser import RawConfigParser
from datetime import datetime, timedelta

import cv2

from Config import Config
from Systems.BaseSystem import BaseSystem
from Utils.Decorators import type_check

logger = logging.getLogger(__name__.split(".")[-1])


class SecuritySystem(BaseSystem):
    """ SecuritySystem class """

    @type_check
    def __init__(self, owner: None) -> None:
        """ Initialize an instance of the SecuritySystem class.

        Arguments:
                owner {MyHome} -- MyHome object which is the owner of the system.
        """

        super().__init__(owner)

        self._owner.event += lambda s, e, d=None: self._onEventReceived(
            s, e, d)

        self.startDelay = timedelta(minutes=15)
        self.sendInterval = timedelta(minutes=5)
        self.numImages = 30

        self._activated = False
        self._startTime = datetime.now() + self.startDelay
        self._prevImages = {}  # camera name / OpenCV image
        self._imageFiles = {}  # camera name / image file name
        self._history = []

    @type_check
    def load(self, configParser: RawConfigParser, data: dict) -> None:
        """ Loads settings and data for the system.

        Arguments:
                configParser {RawConfigParser} -- ConfigParser from which the settings will be loaded.
                data {dict} -- Dictionary from which the system data will be loaded.
        """

        super().load(configParser, data)

        if "history" in data:
            self._history = data["history"]

        self._startTime = datetime.now() + self.startDelay

    @type_check
    def save(self, configParser: RawConfigParser, data: dict) -> None:
        """ Saves settings and data used by the system.

        Arguments:
                configParser {RawConfigParser} -- ConfigParser to which the settings will be saved.
                data {dict} -- Dictionary to which the system data will be saved.
        """

        super().save(configParser, data)

        data["history"] = self._history

    @type_check
    def _onEventReceived(self, sender: object, event: str, _: object) -> None:
        """ Event handler.

        Arguments:
            sender {object} -- Sender of the event.
            event {str} -- Event type.
            data {object} -- Data associated with the event.
        """

        if sender != self or event != "IsEnabledChanged":
            return

        if self._activated:
            self._addHistory("Alarm deactivated")
        self._activated = False
        self._startTime = datetime.now() + self.startDelay
        self._prevImages.clear()
        self._clearImages()

    @type_check
    def update(self) -> None:
        """ Update current system's state. """

        super().update()

        elapsed = datetime.now() - self._startTime
        # send alert
        if elapsed > self.sendInterval:
            self._activated = False
            if len(self._imageFiles) > 0 or self._checkForOfflineCamera():
                self._addHistory("Send alert")
                files = [Config.BinPath + item for _,
                         value in self._imageFiles.items() for item in value]
                if self._owner.sendAlert("Security Alarm Activated!", files, True):
                    self._clearImages()
            else:
                self._addHistory("Skip alert sending")

        # check for motion - if not activated and after delay start
        if not self._activated and elapsed > timedelta():
            self._activated = self._owner.systems["SensorsSystem"].isMotionDetected
            if self._activated:
                logger.info("Alarm Activated")
                self._addHistory("Alarm activated")
                self._owner.event(self, "AlarmActivated")
            self._startTime = datetime.now()

        # get images from cameras and check for movement
        if self._activated:
            for camera in self._owner.systems["SensorsSystem"]._cameras:
                if camera.capture.isOpened():
                    img = camera.getImage()
                    if camera.name not in self._prevImages:  # first image taken
                        self._prevImages[camera.name] = img
                    elif self._findMovement(self._prevImages[camera.name], img):
                        self._saveImage(camera.name, img)
                        self._prevImages[camera.name] = img

    @type_check
    def _saveImage(self, cameraName: str, img: object) -> None:
        """ Save the set image on the disk.

        Arguments:
            cameraName {str} -- Camera name.
            img {object} -- OpenCV image to be saved.
        """

        if cameraName not in self._imageFiles:  # first image
            self._imageFiles[cameraName] = [cameraName + "0.jpg"]
            cv2.imwrite(
                Config.BinPath + self._imageFiles[cameraName][-1], self._prevImages[cameraName])

        self._imageFiles[cameraName].append(
            cameraName + str(len(self._imageFiles[cameraName])) + ".jpg")
        cv2.imwrite(Config.BinPath +
                    self._imageFiles[cameraName][-1], img)

    @type_check
    def _clearImages(self) -> None:
        """ Clear saved cameras images. """

        for _, value in self._imageFiles.items():
            for file in value:
                if os.path.isfile(Config.BinPath + file):
                    os.remove(Config.BinPath + file)
        self._imageFiles.clear()

    @type_check
    def _checkForOfflineCamera(self) -> bool:
        """ Return true if there is a offline camera. """

        for camera in self._owner.systems["SensorsSystem"]._cameras:
            if not camera.capture.isOpened():
                return True
        return False

    @type_check
    def _addHistory(self, entry: str) -> None:
        """ Add entry to history. 

        Arguments:
            entry {str} -- Entry message which will be added.
        """

        self._history.append(f"{datetime.now():%Y-%m-%d %H:%M:%S} {entry}")
        self._history = self._history[-500:]  # keep last 500 entries
        self._owner.systemChanged = True

    @type_check
    def _findMovement(self, image1: object, image2: object) -> bool:
        """ Find movement between two OpenCV images.

        Arguments:
            image1 {object} -- First OpenCV image.
            image2 {object} -- Second OpenCV image.

        Returns:
            bool -- True if there is a movement between the two OpenCV images.
        """

        if image1 is None or image2 is None or \
                image1.shape != image2.shape:
            return False

        diff = cv2.subtract(image2, image1)
        diff = cv2.cvtColor(diff, cv2.COLOR_BGR2GRAY)  # to gray
        _, diff = cv2.threshold(diff, 32, 255, cv2.THRESH_BINARY)  # binarize
        diffValue = diff.mean() 
        if diffValue > 0.01 * 255:
            logger.debug("Camera movement value: %s", diffValue)
            self._addHistory(f"Camera movement value: {diffValue}")
            return True
        return False

File: Systems/SecuritySystem.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;

using MyHome.Models;
using MyHome.Systems.Devices.Sensors;
using MyHome.Utils;

using Newtonsoft.Json;

using NLog;

namespace MyHome.Systems
{
    public class SecuritySystem : BaseSystem
    {
        private static readonly ILogger logger = LogManager.GetCurrentClassLogger();


        [UiProperty(true, "minutes")]
        public int ActivationDelay { get; set; } // minutes

        [UiProperty(true, "minutes")]
        public int SendInterval { get; set; } // minutes

        [UiProperty(true, "percents")]
        public double MovementThreshold { get; set; } // percents

        [UiProperty(true, "minutes")]
        public int PresenceDetectionInterval { get; set; } // minutes

        [UiProperty(true, "person/device IP")]
        public Dictionary<string, string> PresenceDeviceIPs { get; }


        [UiProperty]
        public Dictionary<string, Dictionary<DateTime, string>> History { get; } // roomName / time / action

        [JsonIgnore]
        [UiProperty]
        public List<string> Present { get; private set; } // list of present people


        [JsonIgnore]
        public Dictionary<Room, bool> ActivatedRooms
        {
            get
            {
                lock (this.roomsInfo)
                    return this.roomsInfo.ToDictionary(r => r.Room, r => r.Activated);
            }
        }


        private sealed class RoomInfo
        {
            public Room Room { get; set; }

            public int EnableLevel { get; set; }

            public bool Activated { get; set; }

            public DateTime StartTime { get; set; }

            public List<string> ImageFiles { get; set; }
        }

        [JsonProperty]
        private readonly List<RoomInfo> roomsInfo;

        private readonly Dictionary<string, byte[]> prevImages;

        private DateTime presenceDetectionTimer;


        public SecuritySystem()
        {
            this.ActivationDelay = 15;
            this.SendInterval = 5;
            this.MovementThreshold = 0.02;
            this.PresenceDetectionInterval = 1;
            this.PresenceDeviceIPs = new Dictionary<string, string>();

            this.History = new Dictionary<string, Dictionary<DateTime, string>>();
            this.Present = new List<string>();

            this.roomsInfo = new List<RoomInfo>();
            this.prevImages = new Dictionary<string, byte[]>();
            this.presenceDetectionTimer = DateTime.Now - TimeSpan.FromMinutes(this.PresenceDetectionInterval);

            Directory.CreateDirectory(MyHome.Instance.Config.CameraRecordsPath);
        }

        public void SetEnable(Room room, bool enable, int level = 0)
        {
            lock (this.roomsInfo)
            {
                var roomInfo = this.roomsInfo.Find(r => r.Room == room);
                if (enable)
                {
                    if (roomInfo != null)
                    {
                        logger.Debug($"Try to enable({level}) security alarm for '{room.Name}' room, but it's already enabled({roomInfo.EnableLevel})");
                        if (roomInfo.EnableLevel > level)
                            roomInfo.EnableLevel = level;
                        return;
                    }

                    logger.Info($"Enable({level}) security alarm for '{room.Name}' room");
                    this.roomsInfo.Add(new RoomInfo()
                    {
                        Room = room,
                        EnableLevel = level,
                        Activated = false,
                        StartTime = DateTime.Now + TimeSpan.FromMinutes(this.ActivationDelay),
                        ImageFiles = new List<string>()
                    });
                }
                else
                {
                    if (roomInfo == null)
                    {
                        logger.Debug($"Try to disable security alarm for '{room.Name}' room, but it's not enabled");
                        return;
                    }
                    if (roomInfo.EnableLevel < level)
                    {
                        logger.Debug($"Try to disable({level}) security alarm for '{room.Name}' room, but it is enabled with higher level({roomInfo.EnableLevel})");
                        return;
                    }

                    logger.Info($"Disable({level}) security alarm for '{room.Name}' room");
                    if (roomInfo.Activated)
                        logger.Warn("The alarm is currently activated");
                    else
                        roomInfo.ImageFiles.ForEach(File.Delete);
                    this.roomsInfo.Remove(roomInfo);
                }
                this.SetHistory(room.Name, enable ? "Enabled" : "Disabled");
                MyHome.Instance.SystemChanged = true;
            }
        }

        public void Activate(Room room)
        {
            lock (this.roomsInfo)
            {
                var roomInfo = this.roomsInfo.Find(r => r.Room == room);
                if (roomInfo == null)
                {
                    logger.Trace($"Try to activate security alarm for '{room.Name}' room, but it's not enabled");
                    return;
                }

                if (DateTime.Now - roomInfo.StartTime < TimeSpan.Zero)
                {
                    logger.Debug($"Try to activate security alarm for '{room.Name}' room, but the delay isn't passed");
                    return;
                }

                if (roomInfo.Activated)
                    return;

                roomInfo.Activated = true;
                roomInfo.StartTime = DateTime.Now;
                foreach (var camera in roomInfo.Room.Cameras)
                    this.prevImages.Remove(camera.Room.Name + "." + camera.Name);

                // play security alarm on speakers if no presence detected
                Task.Delay(TimeSpan.FromMinutes(this.PresenceDetectionInterval * 3)).ContinueWith(t =>
                {
                    if (this.roomsInfo.Find(r => r.Room == room)?.Activated == true)
                        MyHome.Instance.PlayAlarm(Devices.Drivers.Types.ISpeakerDriver.AlarmType.Security);
                });

                logger.Info($"Security alarm activated in '{room.Name}' room");
                this.SetHistory(room.Name, "Activated");
                MyHome.Instance.SystemChanged = true;
                MyHome.Instance.Events.Fire(this, GlobalEventTypes.SecurityAlarmActivated, roomInfo.Room);
            }
        }


        protected override void Update()
        {
            base.Update();

            lock (this.roomsInfo)
            {
                foreach (var roomInfo in this.roomsInfo)
                {
                    if (!roomInfo.Activated)
                        continue;

                    var elapsed = DateTime.Now - roomInfo.StartTime;

                    // send alert
                    if (elapsed > TimeSpan.FromMinutes(this.SendInterval))
                    {
                        roomInfo.Activated = false;
                        this.SetHistory(roomInfo.Room.Name, "Enabled");
                        if (roomInfo.ImageFiles.Count > 1 || CheckCameras(roomInfo.Room))
                        {
                            bool sent = Alert.Create($"'{roomInfo.Room.Name}' room security alarm activated!")
                                .Level(Alert.AlertLevel.Critical)
                                .Filenames(roomInfo.ImageFiles)
                                .Send();
                            if (sent)
                            {
                                roomInfo.ImageFiles.ForEach(File.Delete);
                                roomInfo.ImageFiles.Clear();
                            }
                        }
                        else
                        {
                            logger.Info($"Skip alert sending for '{roomInfo.Room.Name}' room");
                            roomInfo.ImageFiles.ForEach(File.Delete);
                            roomInfo.ImageFiles.Clear();
                        }

                        MyHome.Instance.SystemChanged = true;
                    }

                    // use task since Camera getImage can block
                    Task.Run(() =>
                    {
                        foreach (var camera in roomInfo.Room.Cameras)
                        {
                            if (camera.IsOpened()) // try to open the camera and if succeed
                                this.SaveImage(camera, roomInfo);
                        }
                    });
                }
            }

            // presence detection
            if (DateTime.Now - this.presenceDetectionTimer > TimeSpan.FromMinutes(this.PresenceDetectionInterval))
            {
                this.presenceDetectionTimer = DateTime.Now;
                Task.Run(() =>
                {
                    var newPresence = this.DetectPresence().OrderBy(x => x);

                    if (!newPresence.SequenceEqual(this.Present))
                    {
                        foreach (var name in this.PresenceDeviceIPs.Keys)
                        {
                            if (newPresence.Contains(name) && !this.Present.Contains(name)) // new present
                                this.SetHistory(name, "Present");
                            else if (!newPresence.Contains(name) && this.Present.Contains(name)) // someone left
                                this.SetHistory(name, "Left");
                        }

                        if (newPresence.Any() != this.Present.Any()) // if there is a change in presence at all, not per person 
                            MyHome.Instance.Events.Fire(this, GlobalEventTypes.PresenceChanged, newPresence.Any());

                        logger.Info(newPresence.Any() ? "Presence change detected: " + string.Join(", ", newPresence) : "No presence");
                        this.Present = newPresence.ToList();
                    }
                });
            }
        }


        private void SaveImage(Camera camera, RoomInfo roomInfo)
        {
            var image = camera.GetImage(false);
            if (image == null)
                return;
            if (this.prevImages.ContainsKey(camera.Room.Name + "." + camera.Name))
            {
                // if no movement between current and previous image - skip saving
                var diffPercent = camera.DiffImages(this.prevImages[camera.Room.Name + "." + camera.Name], image);
                if (diffPercent < this.MovementThreshold)
                    return;
            }
            this.prevImages[camera.Room.Name + "." + camera.Name] = image;

            var filename = $"{camera.Room.Name}_{camera.Name}_{DateTime.Now:yyyy-MM-dd_HH-mm-ss}.jpg";
            var filepath = Path.Combine(MyHome.Instance.Config.CameraRecordsPath, filename);

            try
            {
                File.WriteAllBytes(filepath, image);
                roomInfo.ImageFiles.Add(filepath);
            }
            catch { /* nothing to do */ }
        }

        private List<string> DetectPresence()
        {
            var result = new List<string>();
            using var ping = new System.Net.NetworkInformation.Ping();
            // retry 5 times
            for (int i = 0; i < 5; i++)
            {
                foreach (var kvp in this.PresenceDeviceIPs)
                {
                    if (result.Contains(kvp.Key))
                        continue;
                    try
                    {
                        var reply = ping.Send(System.Net.IPAddress.Parse(kvp.Value), 1000);
                        logger.Trace($"Ping: {kvp.Value}, {reply.RoundtripTime}, {reply.Status}");
                        if (reply.Status == System.Net.NetworkInformation.IPStatus.Success)
                            result.Add(kvp.Key);
                        if (result.Count == this.PresenceDeviceIPs.Count) // all devices are present
                            return result;
                    }
                    catch (Exception e)
                    {
                        if (i == 0)
                        {
                            logger.Error($"Failed to detect presence for {kvp.Key} ({kvp.Value})");
                            logger.Debug(e);
                        }
                    }
                }
            }
            return result;
        }

        private void SetHistory(string actor, string action)
        {
            if (!this.History.ContainsKey(actor))
                this.History.Add(actor, new Dictionary<DateTime, string>());

            this.History[actor][DateTime.Now] = action;

            // remove entries older than one week
            var weekAgo = DateTime.Now.AddDays(-7);
            foreach (var key in this.History.Keys)
            {
                foreach (var time in this.History[key].Keys)
                {
                    if (time < weekAgo)
                        this.History[key].Remove(time);
                }
            }
        }


        private static bool CheckCameras(Room room)
        {
            // it there is no cameras or there is offline camera
            return !room.Cameras.Any() || room.Cameras.Any(camera => !camera.IsOpened());
        }
    }
}
